### PR TITLE
Remove IterDataPipe from Inference pipeline

### PR DESCRIPTION
**Pull request by gitttt-1234**
This PR removes the Iterdatapipe dependency from the inference pipeline. Currently, we use IterDatapipe modules for creating the data pipeline from labels while running inference on `.slp` files. Since we're moving to Litdata (Related issue: #80), we're removing the IterDatapipe modules and implementing a `thread-based` approach, similar to our current `VideoReader` (Related issue: #26). 

## Release Notes

- **New Features**
	- Introduced `LabelsReaderDP` for enhanced data handling in pipelines.
	- Added support for loading pre-trained model weights in `ModelTrainer` classes.
	- New attributes in `CentroidCrop` class to utilize ground-truth centroids during inference.

- **Bug Fixes**
	- Adjusted expected output shapes in various tests to align with updated processing logic.

- **Documentation**
	- Updated docstrings and comments to reflect changes in data provider and processing methods.

- **Tests**
	- Refactored tests to use `LabelsReaderDP` and updated assertions based on new data handling logic.

### CODE REVIEW

File: sleap_nn/data/providers.py [hunk 1]
```diff
@@ -265,15 +266,129 @@ def run(self):
...
+                {
+                    "image": None,
+                    "frame_idx": None,
+                    "video_idx": None,
+                    "orig_size": None,
+                }
+            )
+
+
+class LabelReader(Thread):
```
Comment (talmo) [inline sleap_nn/data/providers.py:291]:
```suggestion
class LabelsReader(Thread):
```

And rename the old one: `LabelsReaderDP` so we don't have to remove it right away.

File: sleap_nn/inference/predictors.py [hunk 2]
```diff
@@ -233,63 +244,64 @@ def _predict_generator(self) -> Iterator[Dict[str, np.ndarray]]:
...
+                    if scale != 1.0:
+                        if self.instances_key:
+                            ex["image"], ex["instances"] = apply_resizer(
+                                ex["image"], ex["instances"]
+                            )
+                        else:
+                            ex["image"] = resize_image(ex["image"], scale)
+                    ex["image"] = apply_pad_to_stride(
+                        ex["image"], self.preprocess_config["max_stride"]
+                    )
```
Comment (talmo) [inline sleap_nn/inference/predictors.py:298]: Move this to a separate `preprocess()` method and note in the docstring that a TODO is to move this into the inference model itself so these ops are exportable.

File: sleap_nn/inference/predictors.py [hunk 3]
```diff
@@ -497,7 +522,7 @@ def from_trained_models(
             An instance of `TopDownPredictor` with the loaded models.
 
             One of the two models can be left as `None` to perform inference with ground
-            truth data. This will only work with `LabelsReader` as the provider.
+            truth data. This will only work with `LabelReader` as the provider.
```
Comment (talmo) [inline sleap_nn/inference/predictors.py:525]:
```suggestion
            truth data. This will only work with `LabelsReader` as the provider.
```

File: sleap_nn/inference/predictors.py [hunk 4]
```diff
@@ -554,86 +579,57 @@ def from_trained_models(
...
+        queue_maxsize: int = 8,
+        video_start_idx=None,
+        video_end_idx=None,
+    ):
         """Make a data loading pipeline.
 
         Args:
             provider: (str) Provider class to read the input sleap files.
-                Either "LabelsReader" or "VideoReader".
+                Either "LabelReader" or "VideoReader".
```
Comment (talmo) [inline sleap_nn/inference/predictors.py:594]:
```suggestion
                Either "LabelsReader" or "VideoReader".
```

File: sleap_nn/inference/predictors.py [hunk 5]
```diff
@@ -554,86 +579,57 @@ def from_trained_models(
...
-                    pipeline,
-                    anchor_ind=self.confmap_config.model_config.head_configs.centered_instance.confmaps.anchor_part,
-                )
-                pipeline = InstanceCropper(
-                    pipeline,
-                    crop_hw=self.data_config.crop_hw,
-                )
+        # LabelReader provider
+        if self.provider == "LabelReader":
+            provider = LabelReader
```
Comment (talmo) [inline sleap_nn/inference/predictors.py:608]:
```suggestion
        if self.provider == "LabelsReader":
            provider = LabelsReader
```

File: sleap_nn/inference/predictors.py [hunk 6]
```diff
@@ -654,20 +650,17 @@ class (doesn't return a pipeline) and the Thread is started in
...
+                queue_maxsize=queue_maxsize,
+                start_idx=video_start_idx,
+                end_idx=video_end_idx,
             )
             self.videos = [self.pipeline.video]
 
         else:
             raise Exception(
-                "Provider not recognised. Please use either `LabelsReader` or `VideoReader` as provider"
+                "Provider not recognised. Please use either `LabelReader` or `VideoReader` as provider"
```
Comment (talmo) [inline sleap_nn/inference/predictors.py:663]:
```suggestion
                "Provider not recognised. Please use either `LabelsReader` or `VideoReader` as provider"
```

File: sleap_nn/inference/predictors.py [hunk 7]
```diff
@@ -877,82 +870,75 @@ def from_trained_models(
...
+        queue_maxsize: int = 8,
+        video_start_idx=None,
+        video_end_idx=None,
+    ):
         """Make a data loading pipeline.
 
         Args:
             provider: (str) Provider class to read the input sleap files.
-                Either "LabelsReader" or "VideoReader".
+                Either "LabelReader" or "VideoReader".
```
Comment (talmo) [inline sleap_nn/inference/predictors.py:885]:
```suggestion
                Either "LabelsReader" or "VideoReader".
```

File: sleap_nn/inference/topdown.py [hunk 8]
```diff
@@ -47,12 +48,17 @@ class CentroidCrop(L.LightningModule):
...
+        anchor_ind: The index of the node to use as the anchor for the centroid. If not
+            provided or if not present in the instance, the midpoint of the bounding box
+            is used instead.
 
     """
 
     def __init__(
         self,
-        torch_model: L.LightningModule,
+        torch_model: L.LightningModule = None,
```
Comment (talmo) [inline sleap_nn/inference/topdown.py:61]:
```suggestion
        torch_model: L.LightningModule | None = None,
```

File: sleap_nn/inference/topdown.py [hunk 9]
```diff
@@ -63,6 +69,8 @@ def __init__(
         crop_hw: tuple = (160, 160),
         input_scale: float = 1.0,
         max_stride: int = 1,
+        use_gt_centroids: bool = False,
+        anchor_ind: int = None,
```
Comment (talmo) [inline sleap_nn/inference/topdown.py:73]:
```suggestion
        anchor_ind: int | None = None,
```

File: sleap_nn/inference/topdown.py [hunk 10]
```diff
@@ -469,38 +515,20 @@ def forward(self, batch: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
             `"pred_peak_vals": (batch_size, n_nodes)`: Confidence
                 values for the instance skeleton points.
         """
-        batch_size = batch["video_idx"].shape[0]
+        if isinstance(self.instance_peaks, FindInstancePeaksGroundTruth):
+            if "instances" not in batch:
+                raise ValueError(
+                    "Ground truth data was not detected... "
+                    "Please load both models when predicting on non-ground-truth data."
+                )
```
Comment (talmo) [inline sleap_nn/inference/topdown.py:523]: Let's have an issue that describes the case where we have a single-node skeleton and we're doing centroid-only inference?

**Review by talmo — CHANGES_REQUESTED** [review]

File: sleap_nn/data/resizing.py [hunk 11]
```diff
@@ -4,7 +4,7 @@
 
 import torch
 import torch.nn.functional as F
-from sleap_nn.data.providers import LabelsReader, VideoReader
+from sleap_nn.data.providers import LabelsReaderDP, VideoReader
```
Comment (talmo) [inline sleap_nn/data/resizing.py:7]:
```suggestion
from sleap_nn.data.providers import LabelsReader, VideoReader
```

File: sleap_nn/data/resizing.py [hunk 12]
```diff
@@ -230,7 +230,7 @@ class SizeMatcher(IterDataPipe):
     def __init__(
         self,
         source_datapipe: IterDataPipe,
-        provider: Optional[Union[LabelsReader, VideoReader]] = None,
+        provider: Optional[Union[LabelsReaderDP, VideoReader]] = None,
```
Comment (talmo) [inline sleap_nn/data/resizing.py:233]:
```suggestion
        provider: Optional[Union[LabelsReader, VideoReader]] = None,
```

**Review by talmo — APPROVED** [review]
You're still using `LabelsReaderDP` in a bunch of places in the test, but I think we want to test the new `LabelsReader`, no?
